### PR TITLE
#443/Replace default numeric input spinners with suitable alternative

### LINKED ISSUE #443: Allow for fine-grained values in numeric inputs while also preserving step size
# 🐛 Bug Report

Some numeric boxes:
 - have step size defined (a value by which value will be incremented/decremented when clicling arrows)
 - allow for finer values to be valid

For example, for the mitigation strength, we want arrows to increment by 0.1, but we allow values like 0.48 too. Currently this does not work, you can have either one or another:

Screenshot from 2020-04-06 22-46-51

## How to reproduce

Steps to reproduce the issue:

 1. Open the application in a browser
 2. Increment-decrement mitigation strength at will
 3. Try to input 0.48, observe the built--in browser validation error

Note, this issue is applied to all of the numeric fields, not only mitigation strength.

## 😯 Current Behavior

Fine-grained values are not accepted

## 🤔 Expected Behavior

Both, step and fin-grained values work

## 💁 Possible Solution

This is related to the validation attributes in `<input />` tags

There are many libraries exist for numeric inputs, however we would prefer a solution that:
 - has all the features that are currently present
 - uses `<input />`
 - looks native for the browser and platform
 - allows for keyboard navigation (e.g. tab focus)
 - accessible for screen readers (not a "build everything from divs" or at least with aria attributes)
 - accessible for navigation
 - can be internationalized

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Browser          |
| Operating System |
|                  |

## Related

 -

### PR DESCRIPTION
**Pull request by abrie**
## Related issues and PRs

Fixes #443, and expands scope to _all_ numeric interval spinners.

## Description

Replaces interval value spinners with a custom component that supports fine-grained values while maintaining step size.

~~Currently using [react-numeric-input](https://github.com/vlad-ignatov/react-numeric-input).~~

Switched to a custom component after several unsuccessful trials of third party components. https://github.com/neherlab/covid19_scenarios/issues/443#issuecomment-610063175

## Impacted Areas in the application

All ~~Mitigation interval's~~ numeric input spinners.

## Testing

Check all spinner-controlled fields on multiple browsers...

### CODE REVIEW

**Comment by nnoll** [comment]
Stylistically I find it odd that the arrows on the mitigation spinner are different than all other spinners. Shouldn't we just replace all spinners with this variant?

**Comment by abrie** [comment]
@nnoll Yup, will do.

File: src/components/Main/Mitigation/MitigationTable.tsx [hunk 1]
```diff
@@ -117,44 +112,45 @@ function MitigationIntervalComponent({
...
-            <FaTrash />
-          </Button>
-        </div>
-      </div>
-      <div className="w-100">
-        {nameError && <p className="my-0 text-right text-danger">{`${t('Intervention name')}: ${nameError}`}</p>}
-        {valueError && <p className="my-0 text-right text-danger">{`${t('Mitigation strength')}: ${valueError}`}</p>}
-      </div>
+        </Col>
+        <Col xs="12" sm="12" md="12" lg="12" xl="12" className="align-self-center pl-1 pr-1 pb-1">
```
Comment (gj262) [inline src/components/Main/Mitigation/MitigationTable.tsx:137]: Hi @abrie, trying to do fine grained responsive layout with bootstrap is a non starter IMO. So I cannot really help you here. If you want to simply replace the `mitigationValue` `FastField` with a `NumericInput` I can give some input make the layout work.

File: src/components/Form/FormNumericInput.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  errors,
+  touched,
+}: FormNumericInputProps<T>) {
+  const isTouched = _.get(touched, identifier)
+  const errorMessage = _.get(errors, identifier)
+  const showError = errorMessage && isTouched
+
+  return (
+    <Field>
+      {() => (
```
Comment (gj262) [inline src/components/Form/FormNumericInput.tsx:36]: Can this `<Field />` be removed? NumericInput does its own Formik `<Field />`

Comment (abrie) [reply src/components/Form/FormNumericInput.tsx]: Yeah `<Field />` doesn't belong here. Removed it.

File: src/components/Main/Mitigation/MitigationTable.tsx (hunk 1 above)
Comment (abrie) [reply src/components/Main/Mitigation/MitigationTable.tsx]: I've reverted the new layout to your version on master. I'm having a lot of trouble keeping track of how stuff works in there; some help integrating a NumericInput would be much appreciated.

**Comment by gj262** [comment]
Hi @abrie. I pushed a change that uses the NumericInput in the MitigationTable and styles it similar to the current table. (I had hoped to do it as a PR to your PR so you could review it first but github had other ideas. Sorry about that.)

**Comment by abrie** [comment]
@gj262 Thanks for the help!

> Hi @abrie. I pushed a change that uses the NumericInput in the MitigationTable and styles it similar to the current table. (I had hoped to do it as a PR to your PR so you could review it first but github had other ideas. Sorry about that.)

**Comment by abrie** [comment]
😳

**Comment by abrie** [comment]
Ok to close?

**Comment by ivan-aksamentov** [comment]
@abrie Always sad to close good eforts... Thanks for your work anyways.

**Pull request closed without merging**